### PR TITLE
Simultaneous support for -01 and -02 of IETF draft

### PR DESCRIPTION
**Pull request by Nealsoni00**
This improves commit 09f105d53afff61249bb781794615e1e909e2125 to support both draft v01 and draft v02 versions of the ietf upload protocol

Addresses https://github.com/tus/tusd/issues/1065

This is important as iOS 17.2 supports only `draft v01` which means that draft is here to stay

### CODE REVIEW

File: pkg/handler/unrouted_handler.go [hunk 1]
```diff
@@ -650,14 +662,24 @@ func (handler *UnroutedHandler) HeadFile(w http.ResponseWriter, r *http.Request)
...
+			} else {
+				resp.Header["Upload-Complete"] = "?0"
+			}
+		} else if currentUploadDraftInteropVersion == Version3 {
+			if uploadComplete {
+				// Upload is complete if we know the size and it matches the offset.
+				resp.Header["Upload-Incomplete"] = "?0"
+			} else {
+				resp.Header["Upload-Incomplete"] = "?1"
+			}
```
Comment (Acconut) [inline pkg/handler/unrouted_handler.go:679]: The logic of setting Upload-Complete or Upload-Incomplete based on the interop version can be refactored into its own function. That should make the code a bit easier to read and avoid duplication.

File: pkg/handler/unrouted_handler.go [hunk 2]
```diff
@@ -17,7 +17,13 @@ import (
 )
 
 const UploadLengthDeferred = "1"
-const currentUploadDraftInteropVersion = "4"
+
+type DraftVersion string
+
+const (
+	Version3 DraftVersion = "3"
+	Version4 DraftVersion = "4"
```
Comment (Acconut) [inline pkg/handler/unrouted_handler.go:25]:
```suggestion
	interopVersion3 DraftVersion = "3" // From draft version -01
	interopVersion4 DraftVersion = "4" // From draft version -02
```

No need to export these constants.

File: pkg/handler/unrouted_handler.go [hunk 3]
```diff
@@ -1337,7 +1366,19 @@ func (handler *UnroutedHandler) lockUpload(c *httpContext, id string) (Lock, err
 // isResumableUploadDraftRequest returns whether a HTTP request includes a sign that it is
 // related to resumable upload draft from IETF (instead of tus v1)
 func (handler UnroutedHandler) isResumableUploadDraftRequest(r *http.Request) bool {
-	return handler.config.EnableExperimentalProtocol && r.Header.Get("Upload-Draft-Interop-Version") == currentUploadDraftInteropVersion
+	interopVersionHeader := handler.getResumeableUploadDraftVersion(r)
+	supportedInteropVersion := interopVersionHeader == Version3 || interopVersionHeader == Version4
+	return handler.config.EnableExperimentalProtocol && supportedInteropVersion
+}
+
+func (handler UnroutedHandler) getResumeableUploadDraftVersion(r *http.Request) DraftVersion {
```
Comment (Acconut) [inline pkg/handler/unrouted_handler.go:1374]:
```suggestion
func getResumableUploadDraftVersion(r *http.Request) DraftVersion {
```

This is not need to be a method on the UnroutedHandler struct.

**Review by Acconut — DISMISSED** [review]
Thank you very much for working on this! I looks great overall but there are a few improvements to make the code easier to read and maintain :)

**Comment by Nealsoni00** [comment]
Appreciate the review, will work to resolve all comments soon!

File: pkg/handler/unrouted_handler.go (hunk 1 above)
Comment (Nealsoni00) [reply pkg/handler/unrouted_handler.go]: I'm finding it difficult to make a version of this that is cleaner without introducing indirection that is hard to follow. I made some minor improvements (and deduplicated code elsewhere) — would love your thoughts.

**Comment by Acconut** [comment]
Let me know once this is ready for another review.

**Comment by Nealsoni00** [comment]
Hey @Acconut it is now ready! lmk if you don't like the style i used to dedulicate the test logic. The goal was to find what was repeated (and made sense to consolidate) and make `for _, test := range experimentalUploadTests {` to run the tests with the same setup code just different input and output.

File: pkg/handler/unrouted_handler.go [hunk 4]
```diff
@@ -9,7 +9,6 @@ import (
 	"mime"
 	"net/http"
 	"regexp"
-	"slices"
```
[the remark this answers is not in the document]
Comment (Nealsoni00) [reply pkg/handler/unrouted_handler.go]: Ahh i removed it because it failed the old go tests. Honestly, it's simpler/cleaner to just do a switch statement.

File: pkg/handler/unrouted_handler.go (hunk 1 above)
Comment (Acconut) [reply pkg/handler/unrouted_handler.go]: No worries, I just did it on my own. Have a look at https://github.com/tus/tusd/pull/1066/commits/1e4cd41656da7c0925ec8f2ebd6d7d94fe5a432b to see what I meant.

**Review by Acconut — APPROVED** [review]
Thank you very much for your work on this, @Nealsoni00! Apologies for the delay on my side. I just added the final touches on my own and now this is ready to be released. I will cut a new release right afterwards as well.

**Comment by Nealsoni00** [comment]
Hey — appreciate you looking into this and getting it merged!
